src/Commands/core/ArchiveRestoreCommands.php

```diff
@@ -0,0 +1,489 @@
+        );
+        $process->mustRun();
+
+        return $this->siteStatus = $process->getOutputAsJson();
+    }
+
+    /**
+     * Copies files from the source to the destination.
+     *
+     * @param string $source
+     *   The source path.
+     * @param string $destination
+     *   The destination path.
+     *
+     * @throws \Exception
+     */
+    protected function rsyncFiles(string $source, string $destination): void
+    {
+        $source = rtrim($source, DIRECTORY_SEPARATOR) . DIRECTORY_SEPARATOR;
+        $destination = rtrim($destination, DIRECTORY_SEPARATOR) . DIRECTORY_SEPARATOR;
+
+        if (
+            !$this->io()->confirm(
+                dt(
+                    'Are you sure you want to sync files from "!source" to "!destination"?',
+                    [
+                            '!source' => $source,
+                            '!destination' => $destination,
+                        ]
+                )
+            )
+        ) {
+            throw new UserAbortException();
+        }
+
+        $this->logger()->info(
+            dt(
+                'Copying files from "!source" to "!destination"...',
+                [
+                    '!source' => $source,
+                    '!destination' => $destination,
+                ]
+            )
+        );
+
+        $options[] = sprintf("-e 'ssh %s'", $this->getConfig()->get('ssh.options'));
+        $options[] = '-akz';
+        if ($this->output()->isVerbose()) {
+            $options[] = '--stats';
+            $options[] = '--progress';
+            $options[] = '-v';
+        }
+
+        $command = sprintf(
+            'rsync %s %s %s',
+            implode(' ', $options),
+            $source,
+            $destination
+        );
+
+        /** @var \Consolidation\SiteProcess\ProcessBase $process */
+        $process = $this->processManager()->shell($command);
+        $process->run($process->showRealtime());
+        if ($process->isSuccessful()) {
+            return;
+        }
+
+        throw new Exception(
+            dt(
+                'Failed to copy files from !source to !destination: !error',
+                [
+                    '!source' => $source,
+                    '!destination' => $destination,
+                    '!error' => $process->getErrorOutput(),
+                ]
+            )
+        );
+    }
+
+    /**
+     * Imports the database dump to the site.
+     *
+     * @param string $databaseDumpPath
+     *   The path to the database dump file.
+     *
+     * @throws \Drush\Exceptions\UserAbortException
+     * @throws \Exception
+     */
+    protected function importDatabase(string $databaseDumpPath, array $options): void
+    {
+        $this->logger()->info('Importing database...');
+
+        if (!is_file($databaseDumpPath)) {
+            throw new Exception(dt('Database dump file !path not found.', ['!path' => $databaseDumpPath]));
+        }
+
+        $sql = SqlBase::create($options);
```

Comment:
We might not be able to rely on database credentials being available, e.g. the archive:restore destination might be an empty directory, and the archive might not have database credentials.

We should support the Drush 8 archive:restore options related to the database credentials:
```
 --db-prefix                               An optional table prefix to use during restore.                                            
 --db-su                                   Account to use when creating the new database. Optional.                                   
 --db-su-pw                                Password for the "db-su" account. Optional.                                                
 --db-url=<mysql://root:pass@host/db>      A Drupal 6 style database URL indicating the target for database restore. If not provided, 
                                           the archived settings.php is used.                                                         
```
I don't think we necessarily need to support `--db-prefix`, but could, for convenience.